### PR TITLE
[feat]: [CDS-43742]: Azure Artifacts APIs

### PR DESCRIPTION
**Pull request by vtxorxwitty**
## Checklist
- [ ] I've documented the changes in the PR description.
- [ ] I've tested this change either in PR or local environment.
- [ ] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>
  
- Feature build: `trigger feature-build`
- Immutable delegate `trigger publish-delegate`
</details>

<details>
  <summary>PR Check triggers</summary>
  
  You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ti0, ti1`
  
- Compile: `trigger compile`
- CodeFormat: `trigger codeformat`
- MessageMetadata: `trigger messagecheck`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- runDockerizationCheck: `trigger dockerizationcheck`
- runAuthorCheck: `trigger authorcheck`
- Checkstyle: `trigger checkstyle`
- PMD: `trigger pmd`
- TI-bootstrap: `trigger ti0`
- TI-bootstrap1: `trigger ti1`
- TI-bootstrap2: `trigger ti2`
- TI-bootstrap3: `trigger ti3`
- TI-bootstrap4: `trigger ti4`
- FunctionalTest1: `trigger ft1`
- FunctionalTest2: `trigger ft2`
- CodeBaseHash: `trigger codebasehash`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/37877)

### CODE REVIEW

**Comment by diptiman-adak** [comment]
Quality gate Quality gate: **_Failed_** 
 


            
 #### 72 Issues 
* Code issue 69 Code issues 
* Anti-pattern 3 Anti-patterns 

 [Details](https://app.embold.io/organization/gh/harness/repositories/7752dbb3099d2dba3b5d37f5931a5e4c/pull-request/37877)

File: 960-api-services/src/main/java/io/harness/artifacts/azureartifacts/client/AzureArtifactsDevopsRestClient.java [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+
+import io.harness.annotations.dev.OwnedBy;
+import io.harness.artifacts.azureartifacts.beans.NGAzureArtifactsProjects;
+
+import retrofit2.Call;
+import retrofit2.http.GET;
+import retrofit2.http.Header;
+
+@OwnedBy(CDC)
+public interface AzureArtifactsDevopsRestClient {
```
Comment (puthrayaharness) [inline 960-api-services/src/main/java/io/harness/artifacts/azureartifacts/client/AzureArtifactsDevopsRestClient.java:20]: Why can we not use the existing client?

File: 960-api-services/src/main/java/io/harness/artifacts/azureartifacts/client/AzureArtifactsRegistryRestClient.java [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+import retrofit2.http.Path;
+import retrofit2.http.Query;
+
+@OwnedBy(CDC)
+public interface AzureArtifactsRegistryRestClient {
+  /**
+   * List Azure Artifacts Feed
+   */
+  @GET("_apis/packaging/feeds?api-version=5.1-preview.1")
+  Call<NGAzureArtifactsFeeds> listAzureArtifactsFeeds(@Header("Authorization") String authHeader);
```
Comment (puthrayaharness) [inline 960-api-services/src/main/java/io/harness/artifacts/azureartifacts/client/AzureArtifactsRegistryRestClient.java:31]: Why can we not use the existing client?

File: 960-api-services/src/main/java/io/harness/artifacts/azureartifacts/client/AzureArtifactsRegistryRestClient.java [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+import io.harness.artifacts.azureartifacts.beans.NGAzureArtifactsPackages;
+
+import retrofit2.Call;
+import retrofit2.http.GET;
+import retrofit2.http.Header;
+import retrofit2.http.Path;
+import retrofit2.http.Query;
+
+@OwnedBy(CDC)
+public interface AzureArtifactsRegistryRestClient {
```
Comment (puthrayaharness) [inline 960-api-services/src/main/java/io/harness/artifacts/azureartifacts/client/AzureArtifactsRegistryRestClient.java:26]: Please use the same client as that of CG.

File: 960-api-services/src/main/java/io/harness/artifacts/azureartifacts/service/AzureArtifactsRegistryUtils.java [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+
+package io.harness.artifacts.azureartifacts.service;
+
+import io.harness.exception.InvalidArtifactServerException;
+
+import java.net.URI;
+import java.net.URISyntaxException;
+
+public class AzureArtifactsRegistryUtils {
+  public static void validateAzureDevopsUrl(String azureDevopsUrl) {
```
Comment (puthrayaharness) [inline 960-api-services/src/main/java/io/harness/artifacts/azureartifacts/service/AzureArtifactsRegistryUtils.java:16]: Can we not use this same as CG? This does even depend on any credentials

File: 970-api-services-beans/src/main/java/io/harness/artifacts/azureartifacts/beans/NGAzureArtifactsFeeds.java [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
+import java.util.List;
+import lombok.Data;
+
+@OwnedBy(CDC)
+@Data
+@JsonIgnoreProperties(ignoreUnknown = true)
+public class NGAzureArtifactsFeeds {
+  private int count;
+  private List<NGAzureArtifactsFeed> value;
```
Comment (puthrayaharness) [inline 970-api-services-beans/src/main/java/io/harness/artifacts/azureartifacts/beans/NGAzureArtifactsFeeds.java:23]: Why can we not use this same as CG?

File: 970-api-services-beans/src/main/java/io/harness/artifacts/azureartifacts/beans/NGAzureArtifactsPackage.java [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+
+import io.harness.annotations.dev.OwnedBy;
+
+import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
+import lombok.Data;
+
+@OwnedBy(CDC)
+@Data
+@JsonIgnoreProperties(ignoreUnknown = true)
+public class NGAzureArtifactsPackage {
```
Comment (puthrayaharness) [inline 970-api-services-beans/src/main/java/io/harness/artifacts/azureartifacts/beans/NGAzureArtifactsPackage.java:20]: Why can we not use the same as CG?

File: 970-api-services-beans/src/main/java/io/harness/artifacts/azureartifacts/beans/NGAzureArtifactsPackageFileInfo.java [hunk 7]
```diff
@@ -0,0 +1,28 @@
...
+
+import io.harness.annotations.dev.OwnedBy;
+
+import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
+import lombok.Data;
+
+@OwnedBy(CDC)
+@Data
+@JsonIgnoreProperties(ignoreUnknown = true)
+public class NGAzureArtifactsPackageFileInfo {
```
Comment (puthrayaharness) [inline 970-api-services-beans/src/main/java/io/harness/artifacts/azureartifacts/beans/NGAzureArtifactsPackageFileInfo.java:20]: Why not use the same as CG?

File: 970-api-services-beans/src/main/java/io/harness/artifacts/azureartifacts/beans/NGAzureArtifactsPackageVersion.java [hunk 8]
```diff
@@ -0,0 +1,26 @@
...
+import io.harness.annotations.dev.OwnedBy;
+
+import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
+import java.util.List;
+import lombok.Data;
+
+@OwnedBy(CDC)
+@Data
+@JsonIgnoreProperties(ignoreUnknown = true)
+public class NGAzureArtifactsPackageVersion {
```
Comment (puthrayaharness) [inline 970-api-services-beans/src/main/java/io/harness/artifacts/azureartifacts/beans/NGAzureArtifactsPackageVersion.java:21]: Please share with CG file.

File: 970-api-services-beans/src/main/java/io/harness/artifacts/azureartifacts/beans/NGAzureArtifactsPackageVersions.java [hunk 9]
```diff
@@ -0,0 +1,24 @@
...
+import io.harness.annotations.dev.OwnedBy;
+
+import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
+import java.util.List;
+import lombok.Data;
+
+@OwnedBy(CDC)
+@Data
+@JsonIgnoreProperties(ignoreUnknown = true)
+public class NGAzureArtifactsPackageVersions {
```
Comment (puthrayaharness) [inline 970-api-services-beans/src/main/java/io/harness/artifacts/azureartifacts/beans/NGAzureArtifactsPackageVersions.java:21]: Please use this same as CG

File: 970-api-services-beans/src/main/java/io/harness/artifacts/azureartifacts/beans/NGAzureArtifactsPackages.java [hunk 10]
```diff
@@ -0,0 +1,24 @@
...
+import io.harness.annotations.dev.OwnedBy;
+
+import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
+import java.util.List;
+import lombok.Data;
+
+@OwnedBy(CDC)
+@Data
+@JsonIgnoreProperties(ignoreUnknown = true)
+public class NGAzureArtifactsPackages {
```
Comment (puthrayaharness) [inline 970-api-services-beans/src/main/java/io/harness/artifacts/azureartifacts/beans/NGAzureArtifactsPackages.java:21]: Please use same as CG

**Review by puthrayaharness — CHANGES_REQUESTED** [review]

File: 440-connector-nextgen/src/main/java/io/harness/connector/entities/embedded/azureartifacts/AzureArtifactsUsernameToken.java [hunk 11]
```diff
@@ -0,0 +1,23 @@
...
+
+import lombok.Builder;
+import lombok.Value;
+import org.hibernate.validator.constraints.NotEmpty;
+import org.springframework.data.annotation.TypeAlias;
+
+@Value
+@Builder
+@TypeAlias("io.harness.connector.entities.embedded.azureartifacts.AzureArtifactsUsernameToken")
+public class AzureArtifactsUsernameToken {
```
Comment (puthrayaharness) [inline 440-connector-nextgen/src/main/java/io/harness/connector/entities/embedded/azureartifacts/AzureArtifactsUsernameToken.java:18]: Rename to PersonalAccessToken

File: 440-connector-nextgen/src/main/java/io/harness/connector/mappers/azureartifacts/AzureArtifactsDTOToEntity.java [hunk 12]
```diff
@@ -0,0 +1,72 @@
...
+      AzureArtifactsHttpCredentialsDTO httpCredentialsDTO) {
+    final AzureArtifactsHttpAuthenticationType type = httpCredentialsDTO.getType();
+
+    return AzureArtifactsHttpAuthentication.builder().type(type).auth(getHttpAuth(type, httpCredentialsDTO)).build();
+  }
+
+  private AzureArtifactsUsernameToken getHttpAuth(
+      AzureArtifactsHttpAuthenticationType type, AzureArtifactsHttpCredentialsDTO httpCredentialsDTO) {
+    if (type == null) {
+      throw new InvalidRequestException("AzureArtifacts Http Auth Type not found");
```
Comment (puthrayaharness) [inline 440-connector-nextgen/src/main/java/io/harness/connector/mappers/azureartifacts/AzureArtifactsDTOToEntity.java:56]: Change the error message.

File: 440-connector-nextgen/src/main/java/io/harness/connector/mappers/azureartifacts/AzureArtifactsEntityToDTO.java [hunk 13]
```diff
@@ -0,0 +1,81 @@
...
+        final AzureArtifactsUsernameToken usernameToken = (AzureArtifactsUsernameToken) auth;
+
+        usernameTokenDTO = AzureArtifactsUsernameTokenDTO.builder()
+                               .tokenRef(SecretRefHelper.createSecretRef(usernameToken.getTokenRef()))
+                               .build();
+
+        break;
+
+      default:
+        throw new UnknownEnumTypeException("AzureArtifacts Http Auth Type", type.getDisplayName());
```
Comment (puthrayaharness) [inline 440-connector-nextgen/src/main/java/io/harness/connector/mappers/azureartifacts/AzureArtifactsEntityToDTO.java:76]: Change the error message.

**Review by puthrayaharness — CHANGES_REQUESTED** [review]

File: 950-delegate-tasks-beans/src/main/java/io/harness/delegate/task/artifacts/azureartifacts/AzureArtifactsDelegateResponse.java [hunk 14]
```diff
@@ -0,0 +1,76 @@
...
+   */
+  String versionRegex;
+
+  /**
+   * Feed
+   */
+  String feed;
+
+  @Builder
+  public AzureArtifactsDelegateResponse(ArtifactBuildDetailsNG buildDetails, ArtifactSourceType sourceType,
```
Comment (puthrayaharness) [inline 950-delegate-tasks-beans/src/main/java/io/harness/delegate/task/artifacts/azureartifacts/AzureArtifactsDelegateResponse.java:62]: Remove unused constructer params

File: 125-cd-nextgen/src/main/java/io/harness/cdng/artifact/mappers/ArtifactConfigToDelegateReqMapper.java [hunk 15]
```diff
@@ -120,6 +123,31 @@ public GithubPackagesArtifactDelegateRequest getGithubPackagesDelegateRequest(
...
+    }
+
+    String version = artifactConfig.getVersion().getValue();
+
+    if (StringUtils.isBlank(version)) {
+      version = "";
+    }
+
+    // If both version and versionRegex are empty, versionRegex is latest among all versions.
+    if (EmptyPredicate.isEmpty(version) && EmptyPredicate.isEmpty(versionRegex)) {
```
Comment (puthrayaharness) [inline 125-cd-nextgen/src/main/java/io/harness/cdng/artifact/mappers/ArtifactConfigToDelegateReqMapper.java:141]: Fix this. Throw error

File: 120-ng-manager/src/main/java/io/harness/pms/expressions/utils/ImagePullSecretUtils.java [hunk 16]
```diff
@@ -183,6 +191,43 @@ private void getImageDetailsFromS3(
...
+      } else {
+        throw new InvalidRequestException("Please select the Auth type as Username-Token");
+      }
+    }
+
+    if (password == null) {
+      throw new InvalidRequestException("The token is null");
+    }
+
+    imageDetailsBuilder.username(username);
```
Comment (puthrayaharness) [inline 120-ng-manager/src/main/java/io/harness/pms/expressions/utils/ImagePullSecretUtils.java:224]: Remove username

File: 125-cd-nextgen/src/main/java/io/harness/cdng/artifact/resources/azureartifacts/AzureArtifactsResourceServiceImpl.java [hunk 17]
```diff
@@ -0,0 +1,352 @@
...
+      return artifactTaskExecutionResponse.getAzureArtifactsProjects();
+
+    } catch (DelegateServiceDriverException ex) {
+      throw new HintException(
+          String.format(HintException.DELEGATE_NOT_AVAILABLE, DocumentLinksConstants.DELEGATE_INSTALLATION_LINK),
+          new DelegateNotAvailableException(ex.getCause().getMessage(), WingsException.USER));
+
+    } catch (ExplanationException e) {
+      throw new HintException(
+          HintException.HINT_GITHUB_ACCESS_DENIED, new InvalidRequestException(e.getMessage(), USER));
```
Comment (puthrayaharness) [inline 125-cd-nextgen/src/main/java/io/harness/cdng/artifact/resources/azureartifacts/AzureArtifactsResourceServiceImpl.java:207]: Fix message.

File: 125-cd-nextgen/src/main/java/io/harness/cdng/artifact/resources/azureartifacts/AzureArtifactsResourceServiceImpl.java [hunk 18]
```diff
@@ -0,0 +1,352 @@
...
+      return artifactTaskExecutionResponse.getBuildDetails().get(0);
+
+    } catch (DelegateServiceDriverException ex) {
+      throw new HintException(
+          String.format(HintException.DELEGATE_NOT_AVAILABLE, DocumentLinksConstants.DELEGATE_INSTALLATION_LINK),
+          new DelegateNotAvailableException(ex.getCause().getMessage(), WingsException.USER));
+
+    } catch (ExplanationException e) {
+      throw new HintException(
+          HintException.HINT_GITHUB_ACCESS_DENIED, new InvalidRequestException(e.getMessage(), USER));
```
Comment (puthrayaharness) [inline 125-cd-nextgen/src/main/java/io/harness/cdng/artifact/resources/azureartifacts/AzureArtifactsResourceServiceImpl.java:175]: Fix

File: 125-cd-nextgen/src/main/java/io/harness/cdng/artifact/resources/azureartifacts/AzureArtifactsResourceServiceImpl.java [hunk 19]
```diff
@@ -0,0 +1,352 @@
...
+import com.google.inject.Inject;
+import com.google.inject.name.Named;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Optional;
+import javax.annotation.Nonnull;
+import javax.validation.Valid;
+import javax.validation.constraints.NotNull;
+
+public class AzureArtifactsResourceServiceImpl implements AzureArtifactsResourceService {
```
Comment (puthrayaharness) [inline 125-cd-nextgen/src/main/java/io/harness/cdng/artifact/resources/azureartifacts/AzureArtifactsResourceServiceImpl.java:65]: Fix hint Github

Comment (puthrayaharness) [inline 125-cd-nextgen/src/main/java/io/harness/cdng/artifact/resources/azureartifacts/AzureArtifactsResourceServiceImpl.java:65]: Please refactor the code to remove the try-catch block everywher.

File: 960-api-services/src/main/java/io/harness/artifacts/azureartifacts/service/AzureArtifactsRegistryServiceImpl.java [hunk 20]
```diff
@@ -0,0 +1,540 @@
...
+        break;
+      }
+    }
+
+    Response<AzureArtifactsPackages> packagesResponse;
+
+    try {
+      packagesResponse = azureArtifactsRegistryRestClient.listPackages(authHeader, feedId, packageType).execute();
+    } catch (IOException e) {
+      throw new InvalidRequestException("Failed to get builds from regex");
```
Comment (puthrayaharness) [inline 960-api-services/src/main/java/io/harness/artifacts/azureartifacts/service/AzureArtifactsRegistryServiceImpl.java:120]: Replace with hint explnation exception

**Review by puthrayaharness — CHANGES_REQUESTED** [review]

**Review by puthrayaharness — APPROVED** [review]

**Review by puthrayaharness — APPROVED** [review]
> Reviewed 18 of 102 files at r1, 31 of 81 files at r2, 32 of 40 files at r3, 20 of 30 files at r4, 11 of 11 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/37877)* status: :shipit: complete! ✓

**Comment by vtxorxwitty** [comment]
trigger codebasehash

**Review by archit-harness — APPROVED** [review]
> Reviewed 1 of 40 files at r3.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/37877)* status: :shipit: complete! ✓

**Review by prashantsharma04 — APPROVED** [review]
> Reviewed 18 of 102 files at r1, 31 of 81 files at r2, 32 of 40 files at r3, 20 of 30 files at r4, 11 of 11 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/37877)* status: :shipit: complete! ✓

**Review by sahilhindwani-harness — APPROVED** [review]
> Reviewed 1 of 40 files at r3.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/37877)* status: :shipit: complete! ✓

**Pull request merged**